### PR TITLE
Unify all attributes into one; support every item; add `default`

### PR DESCRIPTION
**Pull request by wackbyte**
Adds support for all items and the `default` qualifier, all with one attribute.

### Behavior change

My implementation of the attribute has a slightly different behavior:

```rs
// In the original implementation:
#[fn_qualifiers()] // removes the `const` qualifier!
const fn const_fn_or_is_it() {}

// In this new implementation:
#[qualifiers()] // doesn't touch the item's qualifiers at all
const fn definitely_a_const_fn() {}
```

### Parsing logic

In order to support `const`, `fn`, `static`, and `type` items in all positions, I had to implement custom item types to parse them specially.

```rs
// Example:
#[qualifiers(pub)]
static ITEM_STATIC: u8 = 100;

extern "C" {
    #[qualifiers(pub)]
    static FOREIGN_ITEM_STATIC: u8;
    // If this was parsed as a plain `Item`/`ItemStatic`, it would require the `= <expr>` at the end!
}

// (custom `const`, `fn`, and `type` items are required for `default` support)
```

### Modules

> Modules can't be supported due to hygiene issues.

Is this true? In my testing they seemed to work fine (except for the rust-analyzer issue I noted).

### Fields

I did not implement support for named fields. I have a concept in mind, though:

```rs
#[field_qualifiers(x(pub), y(pub))]
struct InHeader {
    x: String,
    y: u32,
    // TODO: how does this work for tuple structs? `#[field_qualifiers(_0(pub))]`?
}
```

### CODE REVIEW

File: tests/testing.nightly [hunk 1]
```diff
@@ -0,0 +1,154 @@
+//! A test for every supported item.
+//! This test requires the nightly toolchain to compile.
```
Comment (wackbyte) [inline tests/testing.nightly:2]: Note: This file is named the way it is so that `cargo test` ignores it (so you don't need nightly installed to compile)

**Comment by wackbyte** [comment]
I've implemented `#[field_qualifiers]` as I described it above.

**Comment by JohnScience** [comment]
@wackbyte You were given the contributor rights. You can merge the pull request if you want.

**Comment by wackbyte** [comment]
Thanks!

**Comment by ryoqun** [comment]
@JohnScience @wackbyte hey, this pr looks quite impressive. :) when will this be published on crates.io as new ver?

this crate is so useful for our usecase like this: https://github.com/solana-labs/solana/pull/32822#discussion_r1292302244 thanks for the great work.

seems the latest published version v0.1.6 works for now. but I'm planning to heavily use this crate across the codebase.

**Comment by wackbyte** [comment]
@JohnScience would have the ability to do that for you.

**Comment by JohnScience** [comment]
@wackbyte You have been given access on crates.io as well

**Comment by JohnScience** [comment]
@ryoqun Restored the legacy API to avoid breakage and published the 0.2 version on crates.io. Please take a look at the [note on legacy attributes](https://github.com/JohnScience/qualifier_attr#note-on-legacy-attributes).

**Comment by ryoqun** [comment]
> @ryoqun Restored the legacy API to avoid breakage and published the 0.2 version on crates.io. Please take a look at the [note on legacy attributes](https://github.com/JohnScience/qualifier_attr#note-on-legacy-attributes).

really thanks so much (even with polishes with bunch of v0.2.x rels); just created update pr at our side: https://github.com/solana-labs/solana/pull/32838

**Pull request merged**